### DISCUSSION TITLE
Using ImageMagick in a memory efficient manner

### DISCUSSION THREAD

**Discussion by 2246sam20** [ImageMagick/ImageMagick#7655] [Help] [answered]
### ImageMagick version

7.1.1-37

### Operating system, version and so on

11

### Description

Hello!
I am using Image magick as a drawing engine in my application for rendering graphical elements on a canvas and then saving the results aa a png image.
But its using too much memory.
Is there a way to use the Magick library in a memory efficient manner?

**Comment by snibgo** [comment]
Your question is vague. What is "too much memory"? What IM do you use, eg Q-number and HDRI? Could you manage with a lower Q-number and without HDRI? Are you freeing resources when you can?

**Reply by 2246sam20** [reply]
Hello!
The output of the command magick —version displays Q16-HDRI x64.
How to change the q number ?

**Comment by fmw42** [comment]
What was your command line?  Are you using -draw efficiently?

**Reply by 2246sam20** [reply]
Hello i have attached a code snippet of my application. Its one of the methods to render a graphical element to a Magick::Image object.

```
void Renderer: renderPolyLine(const Polyline& polyline, Magick::Image & image) {
    try{
    //LogNessage(std::endl+"Rendering polyline:" + polyline.id ); 
        std::vector<Magick::Coordinate> coordinates;
        std::stringstream ss(polyLine.points);
        std::string point;
        while (std: getline(ss, point, ' ')){
            std::size_t commaPos = point.find(',');
            if (commaPos = std::string::npos){
                double x = std::stod(point.substr(0, commaPos));
                double y = std::stod(point.substr (commaPos + 1));
                coordinates.emplace_back(x, y);
            }
        }
        Magick::DrawablePolyline points(coordinates);
        Magick::DrawableList drawList;
       if (!shape.fill.empty()) {
            Magick::DrawableFillColor fillColor(shape.fill);
            drawList.push_back(fillColor);
        }
        if (!shape.fillopacity.empty()) {
            Magick::DrawableFillOpacity fillOpacity(std::stod(shape.fillOpacity));
            drawList.push_back(fillopacity);
        }
        //pushing other drawable objects to the draw list 
        drawList.push_back(points);
        image.draw(drawList);
    }
    catch (const std::exception& e){
        LogError("Error rendering polyline: " + std::string(e.what()));
    }
}
```

**Comment by snibgo** [comment]
Do you think the code you show is using "too much memory"?

I build IM with `automake`, `autoconfig`, `./configure`, `make` and `make install`. `./configure` has options `--with-quantum-depth=8` (or 16 or 32) and `--enable-hdri`. For the smallest memory usage but lowest precision, use depth 8 and no HDRI.

If you use a different build method, then consult its documentation.

**Reply by 2246sam20** [reply]
I am not sure I am just showing how I am using the Magick library in my application.
Anyways After doing the memory analysis below code snippet was causing too much memory use.
So the memory being used will depend on the width and height = 3840 x 2160 x bytes
Lets take bytes = 16 
Memory =  ~130MB

```
        int width = 3840;
        int height = 2160;
        Magick::Image pageImage(Magick::Geometry(width, height), Magick::Color("transparent"));
       
```

Any suggestions on how to improve this?
P.s - I know width and height can be reduced
But lets say i want to keep them same.

**Comment by snibgo** [comment, marked as answer]
If you use depth 8 and no HDRI, this will need only one byte per channel per pixel.

**Reply by 2246sam20** [reply]
Okay! Will check it.
Thanks for the support!

**Reply by 2246sam20** [reply]
So if we take one byte per channel then buffer size of an object created by below example will be  **width x height x 4**
` Magick::Image pageImage(Magick::Geometry(width, height));`
